Make the app build

Currently, `gradle build` fails:

```stderr
> Task :app:test FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:test'.
> There are test sources present and no filters are applied, but the test task did not discover any tests to execute. This is likely due to a misconfiguration. Please check your test configuration. If this is not a misconfiguration
```

--- a/app/src/main/kotlin/university/innopolis/f/grammar/FList.kt
+++ b/app/src/main/kotlin/university/innopolis/f/grammar/FList.kt
@@ -1,5 +1,6 @@
 package university.innopolis.f.grammar
 
+import university.innopolis.f.lexer.Coordinate
 import university.innopolis.f.lexer.FToken
 import university.innopolis.f.parser.ParseException
 
@@ -13,9 +14,15 @@ value class FList(private val elements: MutableList<FElement>) {
             var res = Pair(firstElemIndex, true)
             while (res.second) {
                 res =
-                    parseElement(allTokens, res.first, self.elements).getOrElse {
-                        return Result.failure(it)
-                    }
+                    parseElement(
+                            allTokens = allTokens,
+                            currentElemIndex = res.first,
+                            buffer = self.elements,
+                            openParCoordinate = allTokens[firstElemIndex - 1].coordinate,
+                        )
+                        .getOrElse {
+                            return Result.failure(it)
+                        }
             }
 
             return Result.success(Pair(self, res.first))
@@ -25,10 +32,11 @@ value class FList(private val elements: MutableList<FElement>) {
             allTokens: List<FToken>,
             currentElemIndex: Int,
             buffer: MutableList<FElement>,
+            openParCoordinate: Coordinate,
         ): Result<Pair<Int, Boolean>> {
             val currentToken = allTokens.getOrNull(currentElemIndex)
             if (currentToken == null) {
-                return Result.failure(ParseException.UnmatchedOpeningParen())
+                return Result.failure(ParseException.UnmatchedOpeningParen(openParCoordinate))
             }
             when (currentToken) {
                 is FToken.OpeningParenthesis -> { // recursion
@@ -50,9 +58,10 @@ value class FList(private val elements: MutableList<FElement>) {
                 }
                 is FToken.Quote -> {
                     val res =
-                        parseElement(allTokens, currentElemIndex + 1, buffer).getOrElse {
-                            return Result.failure(it)
-                        }
+                        parseElement(allTokens, currentElemIndex + 1, buffer, openParCoordinate)
+                            .getOrElse {
+                                return Result.failure(it)
+                            }
                     buffer[buffer.lastIndex] = FElement.Quote(buffer.last())
                     return Result.success(res)
                 }

--- a/app/src/main/kotlin/university/innopolis/f/lexer/Coordinate.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/Coordinate.kt
@@ -1,6 +1,12 @@
 package university.innopolis.f.lexer
 
 data class Coordinate(private var _line: UInt = 1u, private var _column: UInt = 1u) {
+    val line
+        get() = _line
+
+    val column
+        get() = _column
+
     override fun toString(): String = "$_line:$_column"
 
     fun clearColumn() {

--- a/app/src/main/kotlin/university/innopolis/f/lexer/CurrentTokenState.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/CurrentTokenState.kt
@@ -37,7 +37,10 @@ data class CurrentTokenState(private var _rawValue: String = "", val startCoordi
 
         return when {
             _rawValue.matches(integerRegex) -> {
-                FToken.Literal(FLiteral.Integer(FInteger(_rawValue.toBigInteger())), startCoordinate)
+                FToken.Literal(
+                    FLiteral.Integer(FInteger(_rawValue.toBigInteger())),
+                    startCoordinate,
+                )
             }
             _rawValue.matches(realRegex) -> {
                 FToken.Literal(FLiteral.Real(FReal(_rawValue.toBigDecimal())), startCoordinate)

--- a/app/src/main/kotlin/university/innopolis/f/lexer/CurrentTokenState.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/CurrentTokenState.kt
@@ -37,23 +37,23 @@ data class CurrentTokenState(private var _rawValue: String = "", val startCoordi
 
         return when {
             _rawValue.matches(integerRegex) -> {
-                FToken.Literal(FLiteral.Integer(FInteger(_rawValue.toBigInteger())))
+                FToken.Literal(FLiteral.Integer(FInteger(_rawValue.toBigInteger())), startCoordinate)
             }
             _rawValue.matches(realRegex) -> {
-                FToken.Literal(FLiteral.Real(FReal(_rawValue.toBigDecimal())))
+                FToken.Literal(FLiteral.Real(FReal(_rawValue.toBigDecimal())), startCoordinate)
             }
             _rawValue.matches(identifierRegex) -> {
                 when (_rawValue) {
-                    "true" -> FToken.Literal(FLiteral.Boolean(FBoolean(true)))
-                    "false" -> FToken.Literal(FLiteral.Boolean(FBoolean(false)))
-                    "null" -> FToken.Literal(FLiteral.Null)
+                    "true" -> FToken.Literal(FLiteral.Boolean(FBoolean(true)), startCoordinate)
+                    "false" -> FToken.Literal(FLiteral.Boolean(FBoolean(false)), startCoordinate)
+                    "null" -> FToken.Literal(FLiteral.Null, startCoordinate)
                     else -> {
                         val atom = FAtom(_rawValue)
                         val maybeKeyword = FKeyword.fromAtom(atom)
                         if (maybeKeyword == null) {
-                            FToken.Atom(atom)
+                            FToken.Atom(atom, startCoordinate)
                         } else {
-                            FToken.Keyword(maybeKeyword)
+                            FToken.Keyword(maybeKeyword, startCoordinate)
                         }
                     }
                 }

--- a/app/src/main/kotlin/university/innopolis/f/lexer/FToken.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/FToken.kt
@@ -13,7 +13,8 @@ sealed class FToken(open val coordinate: Coordinate) {
         override fun toString() = "ClosingParenthesis"
     }
 
-    data class Keyword(val value: FKeyword, override val coordinate: Coordinate) : FToken(coordinate) {
+    data class Keyword(val value: FKeyword, override val coordinate: Coordinate) :
+        FToken(coordinate) {
         override fun toString() = this.value.toString()
     }
 
@@ -21,7 +22,8 @@ sealed class FToken(open val coordinate: Coordinate) {
         override fun toString() = this.value.toString()
     }
 
-    data class Literal(val value: FLiteral, override val coordinate: Coordinate) : FToken(coordinate) {
+    data class Literal(val value: FLiteral, override val coordinate: Coordinate) :
+        FToken(coordinate) {
         override fun toString() = this.value.toString()
     }
 

--- a/app/src/main/kotlin/university/innopolis/f/lexer/FToken.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/FToken.kt
@@ -4,28 +4,28 @@ import university.innopolis.f.grammar.FAtom
 import university.innopolis.f.grammar.FKeyword
 import university.innopolis.f.grammar.FLiteral
 
-sealed class FToken() {
-    object OpeningParenthesis : FToken() {
+sealed class FToken(open val coordinate: Coordinate) {
+    data class OpeningParenthesis(override val coordinate: Coordinate) : FToken(coordinate) {
         override fun toString() = "OpeningParenthesis"
     }
 
-    object ClosingParenthesis : FToken() {
+    data class ClosingParenthesis(override val coordinate: Coordinate) : FToken(coordinate) {
         override fun toString() = "ClosingParenthesis"
     }
 
-    data class Keyword(val value: FKeyword) : FToken() {
+    data class Keyword(val value: FKeyword, override val coordinate: Coordinate) : FToken(coordinate) {
         override fun toString() = this.value.toString()
     }
 
-    data class Atom(val value: FAtom) : FToken() {
+    data class Atom(val value: FAtom, override val coordinate: Coordinate) : FToken(coordinate) {
         override fun toString() = this.value.toString()
     }
 
-    data class Literal(val value: FLiteral) : FToken() {
+    data class Literal(val value: FLiteral, override val coordinate: Coordinate) : FToken(coordinate) {
         override fun toString() = this.value.toString()
     }
 
-    object Quote : FToken() {
+    data class Quote(override val coordinate: Coordinate) : FToken(coordinate) {
         override fun toString() = "Quote"
     }
 }

--- a/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
@@ -80,23 +80,23 @@ private class Lexer {
         }
     }
 
-    private fun processOpeningParenthesis(currentCoordinate: Coordinate): Result<Unit> {
+    private fun processOpeningParenthesis(startCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
-        this.addToken(FToken.OpeningParenthesis(currentCoordinate))
+        this.addToken(FToken.OpeningParenthesis(startCoordinate))
         currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 
-    private fun processClosingParenthesis(currentCoordinate: Coordinate): Result<Unit> {
+    private fun processClosingParenthesis(startCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
-        this.addToken(FToken.ClosingParenthesis(currentCoordinate))
+        this.addToken(FToken.ClosingParenthesis(startCoordinate))
         currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 
-    private fun processQuote(currentCoordinate: Coordinate): Result<Unit> {
+    private fun processQuote(startCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
-        this.addToken(FToken.Quote(currentCoordinate))
+        this.addToken(FToken.Quote(startCoordinate))
         currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }

--- a/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
@@ -19,17 +19,17 @@ private class Lexer {
             }
             when {
                 char == '(' ->
-                    processOpeningParenthesis().getOrElse {
+                    processOpeningParenthesis(currentCoordinate).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
 
                 char == ')' ->
-                    processClosingParenthesis().getOrElse {
+                    processClosingParenthesis(currentCoordinate).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
 
                 char == '\'' ->
-                    processQuote().getOrElse { this.errors.addError(it as InvalidTokenException) }
+                    processQuote(currentCoordinate).getOrElse { this.errors.addError(it as InvalidTokenException) }
 
                 char == '+' ->
                     processPlus().getOrElse { this.errors.addError(it as InvalidTokenException) }
@@ -78,21 +78,21 @@ private class Lexer {
         }
     }
 
-    private fun processOpeningParenthesis(): Result<Unit> {
+    private fun processOpeningParenthesis(currentCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
-        this.addToken(FToken.OpeningParenthesis)
+        this.addToken(FToken.OpeningParenthesis(currentCoordinate))
         return result
     }
 
-    private fun processClosingParenthesis(): Result<Unit> {
+    private fun processClosingParenthesis(currentCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
-        this.addToken(FToken.ClosingParenthesis)
+        this.addToken(FToken.ClosingParenthesis(currentCoordinate))
         return result
     }
 
-    private fun processQuote(): Result<Unit> {
+    private fun processQuote(currentCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
-        this.addToken(FToken.Quote)
+        this.addToken(FToken.Quote(currentCoordinate))
         return result
     }
 

--- a/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
@@ -83,18 +83,21 @@ private class Lexer {
     private fun processOpeningParenthesis(currentCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
         this.addToken(FToken.OpeningParenthesis(currentCoordinate))
+        currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 
     private fun processClosingParenthesis(currentCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
         this.addToken(FToken.ClosingParenthesis(currentCoordinate))
+        currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 
     private fun processQuote(currentCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
         this.addToken(FToken.Quote(currentCoordinate))
+        currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 

--- a/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
@@ -18,42 +18,49 @@ private class Lexer {
                     currentTokenState.copy(startCoordinate = currentCoordinate.copy())
             }
             when {
-                char == '(' ->
+                char == '(' -> {
+                    currentTokenState =
+                        currentTokenState.copy(startCoordinate = currentCoordinate.copy())
                     processOpeningParenthesis(currentTokenState.startCoordinate).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
-
-                char == ')' ->
+                }
+                char == ')' -> {
+                    currentTokenState =
+                        currentTokenState.copy(startCoordinate = currentCoordinate.copy())
                     processClosingParenthesis(currentTokenState.startCoordinate).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
-
-                char == '\'' ->
+                }
+                char == '\'' -> {
+                    currentTokenState =
+                        currentTokenState.copy(startCoordinate = currentCoordinate.copy())
                     processQuote(currentTokenState.startCoordinate).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
-
-                char == '+' ->
+                }
+                char == '+' -> {
+                    currentTokenState =
+                        currentTokenState.copy(startCoordinate = currentCoordinate.copy())
                     processPlus().getOrElse { this.errors.addError(it as InvalidTokenException) }
-
-                char == '-' ->
+                }
+                char == '-' -> {
+                    currentTokenState =
+                        currentTokenState.copy(startCoordinate = currentCoordinate.copy())
                     processMinus().getOrElse { this.errors.addError(it as InvalidTokenException) }
-
+                }
                 char == '.' ->
                     processDot().getOrElse { this.errors.addError(it as InvalidTokenException) }
-
                 char.isAsciiDigit() -> processDigit(char)
                 char.isAsciiLetter() ->
                     processLetter(char).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
-
                 char.isWhitespace() -> {
                     processWhitespace().getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
                 }
-
                 else ->
                     this.errors.addError(
                         InvalidTokenException.UnsupportedCharacter(
@@ -83,21 +90,18 @@ private class Lexer {
     private fun processOpeningParenthesis(startCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
         this.addToken(FToken.OpeningParenthesis(startCoordinate))
-        currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 
     private fun processClosingParenthesis(startCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
         this.addToken(FToken.ClosingParenthesis(startCoordinate))
-        currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 
     private fun processQuote(startCoordinate: Coordinate): Result<Unit> {
         val result = this.addCurrentTokenIfNotEmpty()
         this.addToken(FToken.Quote(startCoordinate))
-        currentTokenState = currentTokenState.copy(startCoordinate = currentCoordinate.copy())
         return result
     }
 

--- a/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/Lexer.kt
@@ -19,17 +19,19 @@ private class Lexer {
             }
             when {
                 char == '(' ->
-                    processOpeningParenthesis(currentCoordinate).getOrElse {
+                    processOpeningParenthesis(currentTokenState.startCoordinate).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
 
                 char == ')' ->
-                    processClosingParenthesis(currentCoordinate).getOrElse {
+                    processClosingParenthesis(currentTokenState.startCoordinate).getOrElse {
                         this.errors.addError(it as InvalidTokenException)
                     }
 
                 char == '\'' ->
-                    processQuote(currentCoordinate).getOrElse { this.errors.addError(it as InvalidTokenException) }
+                    processQuote(currentTokenState.startCoordinate).getOrElse {
+                        this.errors.addError(it as InvalidTokenException)
+                    }
 
                 char == '+' ->
                     processPlus().getOrElse { this.errors.addError(it as InvalidTokenException) }

--- a/app/src/main/kotlin/university/innopolis/f/lexer/TokenizeException.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/TokenizeException.kt
@@ -24,7 +24,8 @@ sealed class InvalidTokenException(open val coordinate: Coordinate) : IllegalArg
         override fun toString() = "$coordinate: Unexpected character: $c"
     }
 
-    class Build(val token: String, override val coordinate: Coordinate) : InvalidTokenException(coordinate) {
+    class Build(val token: String, override val coordinate: Coordinate) :
+        InvalidTokenException(coordinate) {
         override fun toString() = "$coordinate: Invalid token: $token"
     }
 }

--- a/app/src/main/kotlin/university/innopolis/f/lexer/TokenizeException.kt
+++ b/app/src/main/kotlin/university/innopolis/f/lexer/TokenizeException.kt
@@ -13,18 +13,18 @@ class TokenizeException(
     override fun toString() = this.errors.joinToString("\n") { "- $it" }
 }
 
-sealed class InvalidTokenException(coordinate: Coordinate) : IllegalArgumentException() {
-    class UnsupportedCharacter(val c: Char, val coordinate: Coordinate) :
+sealed class InvalidTokenException(open val coordinate: Coordinate) : IllegalArgumentException() {
+    class UnsupportedCharacter(val c: Char, override val coordinate: Coordinate) :
         InvalidTokenException(coordinate) {
         override fun toString() = "$coordinate: Unsupported character: $c"
     }
 
-    class UnexpectedForm(val c: Char, val coordinate: Coordinate) :
+    class UnexpectedForm(val c: Char, override val coordinate: Coordinate) :
         InvalidTokenException(coordinate) {
         override fun toString() = "$coordinate: Unexpected character: $c"
     }
 
-    class Build(val token: String, val coordinate: Coordinate) : InvalidTokenException(coordinate) {
+    class Build(val token: String, override val coordinate: Coordinate) : InvalidTokenException(coordinate) {
         override fun toString() = "$coordinate: Invalid token: $token"
     }
 }

--- a/app/src/main/kotlin/university/innopolis/f/parser/ParseException.kt
+++ b/app/src/main/kotlin/university/innopolis/f/parser/ParseException.kt
@@ -1,14 +1,15 @@
 package university.innopolis.f.parser
 
+import university.innopolis.f.lexer.Coordinate
 import university.innopolis.f.lexer.TokenizeException
 
 sealed class ParseException() : IllegalArgumentException() {
-    class UnmatchedOpeningParen() : ParseException() {
-        override fun toString() = "Unmatched opening parenthesis"
+    class UnmatchedOpeningParen(val coordinate: Coordinate) : ParseException() {
+        override fun toString() = "$coordinate: Unmatched opening parenthesis"
     }
 
-    class UnmatchedClosingParen() : ParseException() {
-        override fun toString() = "Unmatched closing parenthesis"
+    class UnmatchedClosingParen(val coordinate: Coordinate) : ParseException() {
+        override fun toString() = "$coordinate: Unmatched closing parenthesis"
     }
 
     class Tokenization(val e: TokenizeException) : ParseException() {

--- a/app/src/main/kotlin/university/innopolis/f/parser/Parser.kt
+++ b/app/src/main/kotlin/university/innopolis/f/parser/Parser.kt
@@ -44,7 +44,7 @@ class Parser {
                 return Result.success(nextIndex)
             }
             is FToken.ClosingParenthesis -> {
-                return Result.failure(ParseException.UnmatchedClosingParen())
+                return Result.failure(ParseException.UnmatchedClosingParen(currentToken.coordinate))
             }
             is FToken.Atom -> {
                 elements.add(FElement.Atom(currentToken.value))

--- a/app/src/test/kotlin/university/innopolis/f/AppTest.kt
+++ b/app/src/test/kotlin/university/innopolis/f/AppTest.kt
@@ -1,3 +1,11 @@
 package university.innopolis.f
 
-class AppTest {}
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+
+class AppTest {
+    @Test
+    fun addition_isCorrect() {
+        assertEquals(4, 2 + 2)
+    }
+}

